swarm/src/lib.rs

```diff
@@ -513,24 +512,30 @@ where
 
         let dials = addresses
             .into_iter()
-            .map(|a| match p2p_addr(peer_id, a) {
-                Ok(address) => {
-                    let dial = match dial_opts.role_override() {
-                        Endpoint::Dialer => self.transport.dial(address.clone()),
-                        Endpoint::Listener => self.transport.dial_as_listener(address.clone()),
-                    };
-                    match dial {
-                        Ok(fut) => fut
-                            .map(|r| (address, r.map_err(TransportError::Other)))
-                            .boxed(),
-                        Err(err) => futures::future::ready((address, Err(err))).boxed(),
+            .map(|address| {
+                let address_result = match peer_id {
+                    Some(peer_id) => address.with_p2p(peer_id),
+                    None => Ok(address),
+                };
```

Comment:
I think we can express this neatly using `Option::map_or`.

Reply:
It would indeed be cleaner but I don't think it is possible without a `.clone()`.

If I do the following I get `used of moved value`:
```rust
let address_result = peer_id.map_or(Ok(address), |peer_id| address.with_p2p(peer_id));
```

I can make it work with cloning `address` like this:
```rust
let address_result = peer_id.map_or(Ok(address.clone()), |peer_id| address.with_p2p(peer_id));
```

What do you prefer ? Shorter implementation but with cloning or the current one, a little bit longer but without cloning ?

Reply:
`Multiaddr` has an `Arc` inside so cloning is cheap :)

Reply:
Oh that's right. I'm changing this then.